Bring USB Feature branch into master

c1728p9 opened the pull request:
### Description

Bring feature-hal-spec-usb-device into master.

### Pull request type

    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [x] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

<!--
    Optional
    Request additional reviewers with @username
-->

### Release Notes

This feature brings USB Device support into Mbed OS from a feature branch. Targets that support USB include the LPC1768, ARCH_PRO, K64F, NUCLEO_F207ZG, NUCLEO_F412ZG, DISCO_F413ZH, NUCLEO_F413ZH, NUCLEO_F429ZI, NUCLEO_F446ZE, NUCLEO_F746ZG, NUCLEO_F756ZG, NUCLEO_F767ZI, DISCO_F469NI, DISCO_F746NG, DISCO_F769NI, DISCO_L475VG_IOT01A, DISCO_L476VG, RZ_A1H, GR_LYCHEE, DISCO_L496AG, NUCLEO_L496ZG, NUCLEO_L4R5ZI. Supported USB device classes include USBAudio, USBHID, USBMouse, USBKeyboard, USBMIDI, USBMSD, USBSerial and USBCDC.

c1728p9 committed:
Copy USBDevice from unsupported

Copy the USBDevice code out of unsupported and into a top level USB
folder in preparation for development.

squash

c1728p9 committed:
Add USBPhy, USB HAL and Utility class

Add the USBPhy and USBPhyEvents abstract classes, the HAL header using
this class and the EndpointResolver utility class.

c1728p9 committed:
Add USBDevice test code

Add a USB test and the class USBTester.cpp to go along with it.

c1728p9 committed:
Add USBPhy for the LPC17xx

Move the LPC17xx USB driver files from
mbed-os\features\unsupported\USBDevice\targets\TARGET_NXP
and update them to match the new USBPhy API.

c1728p9 committed:
Add Kinetis USB files

c1728p9 committed:
Update Kinetis USB to match the USBPhy API

Copy the Kinetis USB driver files from
mbed-os\features\unsupported\USBDevice\targets\TARGET_Freescale
and update them to match the new USBPhy API.

c1728p9 committed:
Remove mbed 2 USB tests from Travis

Remove the mbed 2 USB tests since USB support will be part of mbed 5.

c1728p9 committed:
Update USB licenses and format code

Update USB licenses from MIT to Apache 2 and run astyle on the code.

c1728p9 committed:
Update USBDevice and configure to use a USBPhy

Update the USBDevice class API so it matches mbed-os's naming
conventions, has a more robust API and uses USBPhy as its backend.

c1728p9 committed:
Add USB utility classes

Add an allocation free linked list implementation. Additionally add
the class AsyncOp which provides blocking and wakeup funcionality
to simplify making asynchronous operations block.

c1728p9 committed:
Add and assert USB callback completion

Assert that when a callback is completed it is the active callback.

c1728p9 committed:
Update the LPC17XX driver to the new API

Update this driver to match the new zero-copy API.

maciejbocianski committed:
USB generic tests

c1728p9 committed:
Change USBDevice read handling

Remove the USBDevice function read_start and automatically start all
reads internally in USBDevice. This patch also renames the function
read_finish to read.

c1728p9 committed:
Update USBTester for simplified read handling

Remove read_start and replace read_finish with read to match the new
USBDevice API.

c1728p9 committed:
Fix USB init problems

Initialize the _transfer structure so the behavior is the same
regardless of prior memory contents. This fixes a hang during
USB testing when CI flags are used.

c1728p9 committed:
Update stall state on set and clear feature

Call USBDevice::endpoint_stall/unstall in the request to set/clear
halt so the endpoint state is properly updated in the array
_endpoint_info.

c1728p9 committed:
Move USBDevice processing out of user callbacks

Perform processing triggered by user callbacks when USB is being
unlocked rather than synchronously. This prevents recursive callbacks
which reduces stack usage. This also prevents state change inside
the user callback which makes the code easier to reason about.

c1728p9 committed:
Allow transfer to start when configuring USB

Allow reads and writes to be started when entering the configured
state.

c1728p9 committed:
Update USBDevice to reflect the updated USBPhy

Update USBDevice so the USB buffers are user supplied.

c1728p9 committed:
Update USBCDC and USBSerial

Update the USBCDC and USBSerial classes to the new API. This patch also
updates the blocking behavior to unblock at the appropriate times and
to let the processor sleep when blocking rather than busy waiting.

c1728p9 committed:
Explicitly return when endpoint 0 is stalled

Stalling endpoint 0 indicates the end of a control transfer.
Return immediately when this occurs rather than continuing processing.

In particular, this patch prevents status from erroneously being
sent when _request_setup returns false inside _complete_request.

c1728p9 committed:
Add abort API to USBDevice

Add the function endpoint_abort and ensure ongoing transfer are
aborted when an endpoint is removed.

c1728p9 committed:
Update USBTester for the new USB API

Update the class USBTester so it works with the new zero-copy
USBDevice API.

c1728p9 committed:
Call disconnect when uninitializing USB

Disconnect USB when uninitializing USB so it is in a well defined
state.

c1728p9 committed:
Update the Kinetis USB driver to the new API

Update this driver to match the new zero-copy API.

c1728p9 committed:
Update USBPhy to allow zero copy transfers

Update the USBPhy to allow for zero copy USB transfers on devices
which support it. This also aligns more closely with the CMSIS USB
driver model.

c1728p9 committed:
Update USBPhy doxygen

Update USBPhy doxygen to pull in handbook copyedits. Also update the
docs to reflect the enhanced USBPhy API.

fkjagodzinski committed:
USB: EndpointResolver: Add a generic method to get a free endpoint

c1728p9 committed:
Fix Kinetis endpoint initialization

Fully setup endpoint before enabling it in endpoint_add. This prevents
undefined behavior due to previous configuration.

c1728p9 committed:
Move USBPhy and related files to USB directory

Move the USBPhy header files into usb/device/USBPhy. Also move hal
and target USB files since mbed 2 does not compile when these files
are present without the USBPhy headers.

Directory restructure summary:
platform/USBPhy*
    to usb/device/USBPhy/USBPhy*
hal/*
    to usb/device/hal/*
targets/TARGET_Freescale/usb/*
    to usb/device/targets/TARGET_Freescale/*
targets/TARGET_NXP/TARGET_LPC176X/usb/*
    to usb/device/targets/TARGET_NXP/*

c1728p9 committed:
Remove the USB directory from mbed 2 travis

Don't include the USB directory when testing mbed 2 on travis.

c1728p9 committed:
Correct Kinetis 256 byte ctrl transfers

Correctly handle control transfers which are a multiple of 256 bytes
by checking for a non-zero value in the upper byte of size.

c1728p9 committed:
Limit USB control transfer size

Limit the size of control reads to the max packet size of endpoint
zero. This fixes the handling of large transfers.

c1728p9 committed:
Disable endpoints when disconnecting on Kinetis

Even thought the USB phy is disabled during disconnect, endpoints
still nack packets sent. This patch disables all endpoints so after
disconnect no USB traffic will be sent. This allows brief disconnects
of ~200us (or more) to be reliably detected by the host PC.

c1728p9 committed:
Improve LPC1768 USB stability

Make the following changes to improve stability:
-When disconnecting set address to disabled to prevent nacks
-Clear EP_SLOW one and only once for every interrupt where it is set_address
-Disable and clear control endpoint interrupts when disconnecting

c1728p9 committed:
Update USBTester product name

Change the name reported by USBTester from CDC DEVICE to
MBED TEST DEVICE.

c1728p9 committed:
Add pyusb to requirements

Add pyusb to requirements so the greentea USB tests work.

c1728p9 committed:
Add USBPhy template

Add a template for implementing a USBPhy.

c1728p9 committed:
Update the USBHID classes

Update the USB classes USBHID, USBKeyboard, USBMouse and
USBMouseKeyboard from the unsupported folder.

c1728p9 committed:
Skip USB reset tests on OSX

OSX machines don't allow you to reset USB devices. Skip this test on
OSX so testing can still be used.

maciejbocianski committed:
Fix for USB set_clear_feature_test for OSX

c1728p9 committed:
Fix USB tests

Change the direction used in set interface to match the USB
specification. This allows USB tests to pass on OSX.

fkjagodzinski committed:
Tests: USB: Generic: Add basic endpoint tests

c1728p9 committed:
Update the STM32 USB driver to the new API

Take the code from
mbed-os\features\unsupported\USBDevice\targets\TARGET_STM
as a starting point and use it to fill in the USBPhy template for STM32
devices.

maciejbocianski committed:
USBDevice - add support for deconfiguration

maciejbocianski committed:
USBDevice - add multiconfiguration support

maciejbocianski committed:
code formatting fixes

maciejbocianski committed:
USB reset test fix

maciejbocianski committed:
target: arch_pro enable USBDEVICE support

c1728p9 committed:
Add endpoint_remove_all helper function

Add the helper function endpoint_remove_all which removes all added
endpoints. This is useful for class drivers switching between
different USB configurations.

c1728p9 committed:
Prevent Kinetis USB setup packets at wrong time

When disconnecting USB clear all pending interrupts. This ensures that
pending setup packets are dropped rather than being triggered after
disconnect.

maciejbocianski committed:
USBDevice: endpoint_add/read_start functions return value fix

c1728p9 committed:
Fix Kinetis USB disconnect

When disconnecting in the middle of a setup packet USB stops working
even after being re-connected. This is because the setup packet
suspended endpoint events but nothing resumed endpoint events.
This patch adds code to resume endpoint events in the reset handler.

c1728p9 committed:
Make USB disconnect + connect time explicit

Create the define MIN_DISCONNECT_TIME_US to be used as the amount of
time that must pass between connect and disconnect for the host to
reliably detect reconnection. Replace the wait calls delaying for this
value to wait_us to indicate the precision required. Finally, remove
the unnecessary calls to wait_ms in
repeated_construction_destruction_test.

c1728p9 committed:
Explicitly specify USBPhy used for testing

Add a parameter to the USBPhy class so mock or wrapper USBPhy classes
can be used.

c1728p9 committed:
Remove ep0_write from set_address

The USBDevice stack automatically sends control endpoint status
so this does not and should not be done from the call to set_address.

maciejbocianski committed:
Extend USB test class configuration

jamike committed:
Create HAL_PCD_EP_Abort

maciejbocianski committed:
pyusb_basic test formatting fixes

maciejbocianski committed:
pyusb_basic test add helper functions

maciejbocianski committed:
pyusb_basic test update

test code refactoring
align test to use device multiconfiguration

jeromecoutant committed:
NUCLEO_F103RB: replace CLOCK_SOURCE_USB by DEVICE_USBDEVICE

jeromecoutant committed:
STM32F4: replace CLOCK_SOURCE_USB by DEVICE_USBDEVICE

c1728p9 committed:
Remove blocking option from USB connect

Remove the option to block in USBDevice::connect since this
should be handled at a higher level. Also call init to ensure that
the USBDevice has been initalized.

c1728p9 committed:
Remove extra USBDevice constructor

Remove the second USBDevice constructor since it does not provide
any new functionality.

c1728p9 committed:
Cleanup USBSerial

Make the following changes
-Make blocking the first parameter in the constructor
-Add destructor and proper cleanup
-Add ready and wait_ready functions

c1728p9 committed:
Add USBDevice destructor

Add a destructor to USBDevice to ensure that resources have been
properly released. Additionally add an assert in the destructor that
deinit has already been called. If it has not been called then
interrupts can still occur which may cause a crash.

c1728p9 committed:
Fix typo in USBTester.h

Fix the spelling of product_release.

maciejbocianski committed:
Update usb test to connect and block itself

Update the basic usb test so it does blocking itself since blocking
connect was removed from USBDevice.

c1728p9 committed:
Remove unnecessary USB receive buffer

Remove receive buffer from USBPhy_STM32 and instead passed buffers in
directly.

c1728p9 committed:
Fix return code check in endpoint abort

Assert that HAL_PCD_EP_Abort returns HAL_OK rather that something else.

c1728p9 committed:
Only build in USBPhy_STM32 for supported devices

Define USBSTM_HAL_UNSUPPORTED if DEVICE_USBDEVICE is undefined or
defined to 0.

c1728p9 committed:
Update USBAudio

Update the USBAudio class to use the new USB API. This patch also adds
buffering and blocking functionality so it can be used in practice
from thread context.

c1728p9 committed:
Refactor AsyncOp class

Add the OperationList class to simplify managing async operations.
Add a callback to the AsyncOp class so completion can be signaled with
another mechanism. Finally rework the lock handling so AsyncOp::wait
is passed the lock to use rather than taking it in a constructor.

c1728p9 committed:
Update USBHID and USBCDC for new AsyncOp

Update USB classes to use the new AsyncOp API.

maciejbocianski committed:
USB basic test update

disable suspend/resume test
disable sync frame test
improve reset test

maciejbocianski committed:
USB basic test documentation update

c1728p9 committed:
Require USB classes to implement descriptor

Make the function configuration_desc pure virtual inside USBDevice.
This should be a compile time error since no subclass will work
without a configuration descriptor.

c1728p9 committed:
Update the USBMIDI class

Update the USB class USBMIDI from the unsupported folder.

fkjagodzinski committed:
Tests: USB: Generic: Update documentation

c1728p9 committed:
Fix LPC1768 test failures

Reset endpoints when they are removed. This ensures buffers and the
data toggle bit get reset back to zero.

c1728p9 committed:
Implement LPC1768 USB abort function

Implement the USB abort function so interrupts wont fire for a
transfer that has been aborted. The transfer may still take place
but the buffer passed into endpoint_write or endpoint_read will not
be used.

fkjagodzinski committed:
Tests: USB: Update endpoint tests

Fix the host script issues present on Windows machines.
Add 0 B payload size to bulk endpoints test.
Update halt and abort tests according to PR comments.
Add an explicit request to start reading on OUT endpoints.

c1728p9 committed:
Add the TaskQueue abstract interface and friends

Add the classes Task, TaskBase and TaskQueue. TaskQueue queue is an
interface class which can be implemented by anything which can run
code. Task and TaskBase are concrete classes which allow callbacks
to be posted to a TaskQueue to be run.

c1728p9 committed:
Add the PolledQueue implementation of TaskQueue

Add the PolledQueue class which provides a TaskQueue which is run
by calling PolledQueue::process.

c1728p9 committed:
Update USBMSD

Update the USBMSD class for the new API. Add support for passing in
a BlockDevice directly without the need to extend USBMSD.

c1728p9 committed:
Fix unbalanced USBDevice unlock

Remove calls to unlock which are not preceded by calls to lock.
Also move the location of the unlock underflow assert so unbalanced
unlocking during post processing is caught.

c1728p9 committed:
Update USBDevice endpoint checks to fix asserts

Only assert if disabled endpoints are used when USBDevice is
configured. USBDevice can leave the configured state due to a
reset at any time, which disables all endpoints. Because this
can happen at any time, thread processing could be performing
any endpoint operation. The endpoint operation should return
failure and do nothing in this case, rather than asserting
as this is not an application error.

An assert should only be triggered when an invalid endpoint is
used after the use of USBDevice acknoledges the switch to
configured mode by complete_set_configuration.

In specific this PR fixes the assert caused with the following
sequence:

-ISR: OUT event sent
-ISR: USB reset event
-ISR: USB configure request start
-Thread: OUT event processed on thread and next read starts
        ***endpoint is used while disabled causing an invalid assert***
-Thread: reset event processed
-Thread: configure event processed

This patch fixes this problem by making the following changes:
1. Operations done on disabled endpoints only assert when in the configured state
2. Adding and removing endpoints is only allowed when
    the flag _endpoint_add_remove_allowed is set
3. The flag _endpoint_add_remove_allowed is set on the set
    configuration request and cleared if the request is aborted  or
    fails

c1728p9 committed:
Add default values to USBCDC constructor

Add reasonable defaults to the USBCDC constructor so it is easier to
use.

fkjagodzinski committed:
Tests: USB: Add device config files for Zadig

Zadig is a tool used to install generic USB drivers on Windows machines.
These drivers are necessary to run USB device test suite on Windows
hosts.

c1728p9 committed:
Add circular byte buffer using dynamic memory

Add the ByteBuffer class which is similar to the CircularBuffer class
but uses dynamic memory rather than template parameters and is
optimized for byte transfers. This is required for USBAudio which
needs a high performance circular buffer which can be resized at
runtime.

c1728p9 committed:
Fix isochronous endpoints on LPC1768

Perform isochronous endpoint processing from the frame interrupt
rather than the endpoint interrupt. Isochronous endpoints to not
generate interrupts normally and are intended to be handled from
the start of frame interrupt.

fkjagodzinski committed:
Tests: USB: Add data toggle test

c1728p9 committed:
Fix USB on Kinetis devices

Set correct SYSMPU register for proper USB operation.  This bug was
introduced when the SYSMPU register names and defines were updated
in the commit:
"K64F: Updated the SYSMPU SDK driver"
93f8cfed058174b2de468353c173c0d7b9335ea4

c1728p9 committed:
Alternate Kinetis USB stuck sending bug fix

If an IN endpoint is stalled during a transfer then the data being
sent will repeated and flood the USB bus. This patch prevents
endpoints from being stalled in the middle of a transfer by control
requests by keeping USB suspended until the setup phase of the
control request is done.

c1728p9 committed:
Fix Kinetis bug causing USB to get stuck sending

If an IN endpoint is stalled during a transfer by writing to the
USB ENDPOINT register then the data being sent will repeat and flood
the USB bus. This patch prevents the register write from occurring by
instead writing to the buffer descriptor in RAM and letting the USB
hardware handle setting the stall bit.

Note - Control requests on endpoint 0 do still set the STALL bit
directly. This is not a problem since control endpoints cannot be
stalled externally while a transfer is ongoing.

c1728p9 committed:
Fix various USB warnings

Fix the following warnings:
-[Warning] USBTester.cpp@45,0:  #1299-D: members and base-classes will
    be initialized in declaration order, not in member initialisation
    list order
-[Warning] USBTester.h@41,0:  #1300-D: ~USBTester inherits implicit
    virtual
-[Warning] USBAudio.cpp@345,0:  #1035-D: single-precision operand
    implicitly converted to double-precision
-[Warning] USBHID.cpp@29,0:  #1300-D: ~AsyncSend inherits implicit
    virtual
-[Warning] USBHID.cpp@61,0:  #1300-D: ~AsyncRead inherits implicit
    virtual
-[Warning] USBHID.cpp@93,0:  #1300-D: ~AsyncWait inherits implicit
    virtual
-[Warning] EndpointResolver.cpp@125,26: '<<' in boolean context, did
    you mean '<' ? [-Wint-in-bool-context]

c1728p9 committed:
USBMSD fixes

Make the following fixes:
-deinit in destructor to prevent race conditions
-cancel the reset task before calling it since it may be in progress
-wait for tasks to complete without mutex held
-prevent double connect with _init flag

c1728p9 committed:
Remove endpoint parameter from USB callbacks

Remove the endpoint parameter from endpoint callbacks. This
information is redundant because endpoints are known at
construction time because they must be in the configuration
descriptor.

c1728p9 committed:
Update USBMSD to be more consistent with others

Add second constructor and reorder constructor parameters to match
other USB classes. Also remove the ready() function since there
are no calls that can only mbe made from a ready state.

c1728p9 committed:
Revert "Fix Kinetis bug causing USB to get stuck"

Revert the commit "Fix Kinetis bug causing USB to get stuck sending"
since this change causes stalls to be missed sometimes.

c1728p9 committed:
Abort the current USB transfer when stalling

It is undefined behavior if stalling and unstalling clears an ongoing
transfer. Abort any ongoing transfers explicitly when stalling and
unstalling so the behavior is consistent across devices.

c1728p9 committed:
Reset data toggle on Kinetis when unstalling

Data toggle must be reset to DATA0 when an endpoint is unstalled.
This patch makes that change.

c1728p9 committed:
Update test to restart reads after unstall

After an endpoint is unstalled any pending transfers are terminated.
This patch re-starts any reads that were ongoing.

fkjagodzinski committed:
Tests: USB: Update ep callbacks

Endpoint callbacks no longer have endpoint as a param. This update was
introduced in #7267.

c1728p9 committed:
Revert "Create HAL_PCD_EP_Abort"

Revert the patch "Create HAL_PCD_EP_Abort" in preparation for an
alternate fix.

c1728p9 committed:
Alternate implementation of HAL_PCD_EP_Abort

Update the patch "Create HAL_PCD_EP_Abort" to fix bugs.
This patch adds the low level functions USB_EPStopXfer, USB_EPSetNak,
USB_EPClearNak and the high level function HAL_PCD_EP_Abort so that
transfers can be stopped.

The functions USB_EPSetNak and USB_EPClearNak allow nak to be enabled
or disabled for an endpoint, preventing or allowing further transfers.

The function USB_EPStopXfer stops pending reads and writes started by
USB_EPStartXfer along with clearing and masking any interrupts enabled
by USB_EPStartXfer.

The function HAL_PCD_EP_Abort aborts any transfers on the given
endpoint. When this function completes the transfer interrupt
is guarenteed not to fire for this endpoint. Furthermore, the size
of data transferred during an aborted read can be found by calling
the function HAL_PCD_EP_GetRxCount.

Other notes on this Change:

1.
Prior to this patch the interrupt USB_OTG_DOEPINT_EPDISD was not
handled. When an OUT endpoint was disabled this interrupt occurred
causing the CPU to get stuck repeatedly handling this interrupt. This
is because this interrupt was unmasked but nothing cleared this
interrupt. This patch also adds code to handle and clear this
interrupt to prevent a lockup.

2.
Stopping a transfer on an OUT endpoint requires global nak OUT to
be in effect. Even with this being done, having entries in the rx fifo
prevented an OUT endpoint from being disabled. This behavior is not
mentioned in the Reference Manual.

c1728p9 committed:
Implement endpoint_abort for STM32

Make use of the added function HAL_PCD_EP_Abort to implement
endpoint_abort.

c1728p9 committed:
Fix ST test failures due to endpoint reconfiguration

Also always use the max endpoint size for SetTxFifo
since re-configuring fifos when endpoints are added or removed
causes tests to fail.

c1728p9 committed:
Remove OR from ST endpoint activation

When activating an endpoint assign new data rather than ORing
data to it. This ensures that values set from the previous use
do not effect the current configuration.

fkjagodzinski committed:
Tests: USB: Decrease the max data size for iso ep

Although the USB spec sets the upper limit on FS isochronous endpoint
payloads to 1023 B, this value is hard to test in practice. Moreover,
not all the targets Mbed OS supports (like NUCLEO_F207ZG) are able to
handle all the endpoints set to max.

fkjagodzinski committed:
Tests: USB: Update endpoint halt test

Wait for a locally created Timer thread to finish before returning.

fkjagodzinski committed:
Tests: USB: Check ep buffer is released by abort

Validate that endpoint buffer is not used after a transfer has been
aborted.

c1728p9 committed:
Remove HAL_GetTick from ST USBPhy

Remove the function HAL_GetTick from the ST USBPhy driver. This is no
longer needed as that change has been made on master.

fkjagodzinski committed:
Tests: USB: Add tests for CDC class.

fkjagodzinski committed:
Tests: USB: Add tests for Serial class.

fkjagodzinski committed:
Tests: USB: Fix USB reconnect tests for Win hosts

fkjagodzinski committed:
Tests: USB: Use the serial number to find device

Use the USB device SN instead of VID & PID to find the device port name
on the host system.

fkjagodzinski committed:
Tests: USB: Add Serial test for line coding change

fkjagodzinski committed:
Tests: USB: Wait before closing host's port

This delay eliminates the possibility of the device detecting the port
being closed when still waiting for data.

fkjagodzinski committed:
Tests: USB: Wait before transmitting data to host

Despite having ECHO and ECHOCTL POSIX lflags disabled by default by
pyserial, a delay is needed for host to properly handle data received
from USB serial/CDC device.
With no delay host would echo 0x00-0x31 characters prefixed with '^'.

fkjagodzinski committed:
Tests: USB: Correct reconnect delay

Setting to 1 ms to be on the safe side. The USB spec defines this delay
to be at least 200 us.

fkjagodzinski committed:
Tests: USB: Use SN generated by the host machine

fkjagodzinski committed:
Tests: USB: Update code comments

fkjagodzinski committed:
Tests: USB: Fix Windows USB SN handling

fkjagodzinski committed:
Tests: USB: Increas the size of RX CDC test data

This ensures the flow control is tested.

c1728p9 committed:
Tests: USB: Use DTR to make Serial tests reliable

Update the serial host test to set DTR low before opening a serial
port. Set DTR high once the serial object is ready. This ensures that
no data is sent by the device until the host is ready.

c1728p9 committed:
Tests: USB: Remove TX_DELAY_MS from serial test

Remove TX_DELAY_MS and all the places it was called since it is not
needed when DTR flow control is used.

c1728p9 committed:
Tests: USB: Increase serial open retries

Double the number of attempts that are made to open a serial port.
This makes tests run more reliably on Windows.

c1728p9 committed:
Ignore disabled Kinetis USB endpoint interrupts

Ignore interrupts on disabled USB endpoints. This prevents handling
interrupts when in the wrong state.

Prior to this patch when running the serial test on a K64F the assert
on line 908 of USBDevice.cpp would sometimes be triggered. This
assert indicates that an endpoint 0 IN interrupt occurred before
the device was ready.

This occurs during the test_cdc_usb_reconnect test when the host sends
a "Set Control Line State" USB request and the device acknowledges it
just before USB is disconnected.

c1728p9 committed:
Remove USB disconnect delay from Serial test

Remove the USB disconnect delay since there is no reason this is
needed. Failures which occur without this indicate a device or host
problem.

bcostm committed:
STM32F4 USB: add patch in CubeF4 hal driver

bcostm committed:
STM32F4 USB: remove useless macro

bcostm committed:
STM32F4 USB: enable USB Device in targets.json

tests-usb_device-basic are OK on the following platforms:

NUCLEO_F412ZG
NUCLEO_F413ZH
NUCLEO_F429ZI
NUCLEO_F446ZE
DISCO_F413ZH
DISCO_F469NI

c1728p9 committed:
Fix USB traceback during testing when wrong size

In the pyusb_basic host test check the length of control transfers
before comparing contents for equality.

bcostm committed:
STM32L4 USB: add patch in CubeL4 hal driver

bcostm committed:
STM32L4 USB: add pins definition for STM32L496/L4R5 MCUs

fkjagodzinski committed:
Tests: USB: DTR fix for Linux hosts

A DTR line is used to signal that the host has configured a terminal and
is ready to transmit and receive data from the USB CDC/Serial device.
When this test suite is run with the use of a Linux host, a workaround has
to be used to overcome some platform specific DTR line behavior.
Every time the serial port file descriptor is opened, the DTR line is
asserted until the terminal attributes are set.
As a consequence, the device receives a premature DTR signal with a
duration of 200-500 us before the correct, long-lasting DTR signal set by
the host-side test script. (tested on the Linux kernel 4.15.0)

The solution is to wait for the first DTR spike, ignore it, and wait for
the correct DTR signal again.

bcostm committed:
STM32F7 USB: add patch in CubeF7 hal driver

bcostm committed:
STM32F7 USB: change VBUS pin for DISCO_F746NG FS

bcostm committed:
STM32F7 USB: add NUCLEO_F756ZG in pins definition

bcostm committed:
STM32L4 USB: move HAL_PCD_EP_Abort function

This function is for USB_OTG_FS devices only. Move it in the correct place (in "#ifdef USB_OTG_FS" area).

bcostm committed:
STM32F7 USB: enable USBDEVICE in targets.json

fkjagodzinski committed:
Tests: USB: Add a README for Linux users

Explain how to overcome xHCI limitations in kernels prior to v4.17 and
successfully run USB tests.

TomoYamanaka committed:
Renesas : Implement USB Device feature

I implemented USB Device feature for Renesas mbed boards.
The code referenced the following code as a starting point and is implemented it by inheritting USBPhy same as other boards.
(mbed-os\features\unsupported\USBDevice\targets\TARGET_RENESAS)

fkjagodzinski committed:
Tests: USB: Fix 'endpoint halt' test

Abort all endpoint transfers before running the test again.
Use an updated vendor request to explicitly restart device reads.

fkjagodzinski committed:
Tests: USB: Use astyle to format the test class

bcostm committed:
STM32L4 USB: enable USBDEVICE

c1728p9 committed:
Remove inclusion of mbed.h from USB

Remove mbed.h from USB files and fix the build errors this causes.
This is required to pass CI.

c1728p9 committed:
Applied suggested astyle fixes

0xc0170 commented on TESTS/host_tests/usb_device_serial.py:
```diff
@@ -0,0 +1,346 @@
+"""
+mbed SDK
+Copyright (c) 2018 ARM Limited
```
Please add SPDX identifiers for new files

c1728p9 replied:
done

c1728p9 committed:
Applied more astyle fixes

c1728p9 committed:
Update files to include SPDX-License-Identifier

Update the header of all files to use a newer license template which
includes SPDX-License-Identifier.

bridadan commented on TESTS/host_tests/pyusb_basic.py:
```diff
@@ -0,0 +1,1814 @@
+    """
+
+    list = [64, 256]
+    dev = y
```
Holy smokes, what a host test! 🚢

bridadan commented on TESTS/usb_device/basic/README.md:
```diff
@@ -0,0 +1,15 @@
+When you run the `tests-usb_device-basic` test suite on a Linux machine, please make
+sure that at least one of the following prerequisites are met:
+* using the Linux kernel ***4.17* or newer**,
+* using the ***eHCI*** USB driver instead of *xHCI*.
```
@ARMmbed/mbed-os-test FYI

cmonr replied:
@c1728p9 What's the technical reason for this limitation?

Are we expecting devlopers to have to load a specific OS driver to be able to test in Linux?

fkjagodzinski replied:
> @c1728p9 What's the technical reason for this limitation?

Basically, an incomplete implementation of *xHCI* driver in kernels prior to 4.17. Please see the rest of [this readme](https://github.com/ARMmbed/mbed-os/blob/065df0f0e85468feaacaa30b7e3a6e884a35af61/TESTS/usb_device/basic/README.md). I've also included some links for reference.

cmonr replied:
> Basically, an incomplete implementation of xHCI driver in kernels prior to 4.17. Please see the rest of this readme. I've also included some links for reference.

Damn, those are actually good technical reasons...

Thanks for having them in the readme.

bridadan commented on requirements.txt:
```diff
@@ -21,3 +21,4 @@ fuzzywuzzy>=0.11,<=0.17
 pyelftools>=0.24,<=0.25
 git+https://github.com/armmbed/manifest-tool.git@v1.4.6
 icetea>=1.2.1,<1.3
+pyusb>=1.0.0
```
Could you also add an upper bound on the API version?
```suggestion
pyusb>=1.0.0,<2.0.0
```

bridadan reviewed (approved):
Seems ok to me!

Will there be a future PR to remove the unsupported USB stack? https://github.com/ARMmbed/mbed-os/tree/master/features/unsupported/USBDevice

bridadan committed:
Update requirements.txt

Add an upper bound to the API version of pyusb.

Co-Authored-By: c1728p9 <butleja10511@hotmail.com>

fkjagodzinski reviewed (approved):
Looks good for me :+1:
